### ISSUE TITLE
Reusing memory buffer for multiple Halide Funcs

### ISSUE THREAD

**Issue by gchauras** [halide/Halide#606]
This is asking for ideas, not reporting an issue. I need to:
- perform some computation in an update def of Func A
- save a small subset of the the result
- perform another update step on the buffer A overwriting the prev result.

Since I need to save the intermediate result, I have two options:

- Inject another update def between the two which saves out the subset in another part of the same buffer of A. For this, I need a slightly bigger buffer for A but the memory layout gets all messed up hitting performance.

- Perform each update in a different Func. This way I can save the result, but mutliple Func's cause allocation of multiple buffers and copying the whole buffer as many times. Is it possible for multiple buffers to share the same memory, not possible I guess?

Do you have any suggestions? Thanks a lot for all the help on all the other issues I had.

**Comment by dsharletg** [comment]
This is ugly, but if you want to avoid the change in memory layout, you
could store the prefix sums in a few rows at the end of the image, instead
of in each row:

f(x,y) = undef()
f(rx,y) = in(rx,y)   // init buffer
f(rx,y) += f(rx-1, y)  // prefix sum
f(y%WIDTH, HEIGHT + y/WIDTH) = f(WIDTH-1, y)  // save the last element
// repeat prefix sum etc

I can't think of any other much better ways to do this. It would be great
if you could use intermediate stages in other Funcs, but this is a
nightmare to implement for all but the most trivial schedules.

**Comment by gchauras** [comment]
Yeah, that's basically what I am trying to do. Since I have prefix sums in both x and y dimensions, I have to choose between storing to store the last elements at the end of each row, at the end of each column or both. At first I stored x last elements at the end of row and y last elements at the end of columns but this was slow. I am now trying to pack everything in the same chunk (row or column) which seems to be 50% faster. Thanks for the answer. I just wanted to check if there was any other trick I did not know.
